scripts/katana_version/version.py

```diff
@@ -0,0 +1,388 @@
+    )
+    if config.version_from_environment_variable:
+        env_version = config.version_from_environment_variable
+        if env_version.release != computed_version.release:
+            logger.warning(
+                "The KATANA_VERSION environment variable does not match the version in the source code: "
+                f"{env_version} does not match {computed_version}"
+            )
+        return env_version
+    else:
+        return computed_version
+
+
+def git_find_closest_core_branch(config, commit, repo_path):
+    if not config.has_git:
+        return None
+
+    branch_patterns = [
+        f"{config.upstream_remote}/master",
+        f"{config.upstream_remote}/release/v*",
+        f"{config.upstream_remote}/variant/*",
+    ]
+    branches = [
+        b for pat in branch_patterns for b in git.find_branches(pat, repo_path, prefix="remotes", sort="-creatordate")
+    ]
+
+    if not branches:
+        return None
+
+    def branch_ahead_count(branch):
+        return git.get_commit_count(git.merge_base(commit, branch, repo_path), commit, repo_path)
+
+    nearest_branch = min(branches, key=branch_ahead_count)
+    return nearest_branch
+
+
+def get_explicit_version(config, k_commit: str, use_working_copy: bool, katana_repo_path, variant=None, no_dev=False):
+    tag_version = get_tag_version(config, k_commit, katana_repo_path)
+    explicit_version = tag_version or get_config_version(
+        None if use_working_copy else k_commit, katana_repo_path, no_dev=no_dev
+    )
+    if explicit_version.local and variant and variant != explicit_version.local:
+        logger.warning(
+            f"You are overriding the repository variant {explicit_version.local} with build-time variant {variant}."
+        )
+    variant = variant or explicit_version.local
+    return explicit_version, variant
+
+
+def get_config_version(k_commit, katana_repo_path, no_dev=False) -> version.Version:
+    if k_commit:
+        version_str = capture_command(
+            "git", *git.dir_arg(katana_repo_path), "show", "-p", f"{k_commit}:{CONFIG_VERSION_PATH}"
+        )
+    else:
+        with open(katana_repo_path / CONFIG_VERSION_PATH, "rt") as version_file:
+            version_str = version_file.read()
+    ver = version.Version(version_str.strip())
+
+    if no_dev:
+        return ver
+
+    return add_dev_to_version(ver)
+
+
+def get_tag_version(config, commit, repo_path):
+    if not config.has_git or not commit:
+        return None
+    tag_version = None
+    version_tags = [m for m in (VERSION_TAG_RE.match(t) for t in git.get_tags_of(commit, repo_path)) if m]
+    if len(version_tags) > 1:
+        logger.warning(f"There is more than one version tag at the given commit. Picking one arbitrarily.")
+    if version_tags:
+        tag_version = version.Version(version_tags[0].group("version"))
+    return tag_version
+
+
+def compute_commit_count(config, commit, last_version_commit, repo_path, pretend_master, core_branch):
```

Comment:
I suppose I should be happy that this isn't a bash script.

Reply:
Probably. I really overengineered this, but I think it's a lot better than a bash script that tries to do all this.